### PR TITLE
Add node config option for a resolv.conf to read

### PR DESCRIPTION
**Pull request by smarterclayton**
Supports changing nameservers without having to rewrite node config,
separates local config from master config. Intended to allow for cases
where dnsmasq is in front of the node, but upstream config needs to be
generated by network manager.

@sdodson for 3.6, @liggitt if you have a minute to look

### CODE REVIEW

**Comment by smarterclayton** [comment]
[test]

File: pkg/cmd/server/api/types.go [hunk 1]
```diff
@@ -194,6 +194,11 @@ type NodeConfig struct {
 	// system, this value should be set to the upstream nameservers dnsmasq resolves with.
 	DNSNameservers []string
 
+	// DNSUpstreamResolvConf is a path to a resolv.conf file that contains settings for an upstream server.
+	// Only the nameservers and port fields are used. The file must exist and parse correctly. It overrides
+	// nameservers in DNSNameservers if set.
+	DNSUpstreamResolvConf string
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: `Upstream` in the field name is weird...

File: pkg/cmd/server/admin/create_nodeconfig.go [hunk 2]
```diff
@@ -43,6 +43,7 @@ type CreateNodeConfigOptions struct {
 	DNSBindAddress      string
 	DNSDomain           string
 	DNSIP               string
+	DNSUpstreamResolvConf string
```
Comment (liggitt) [inline pkg/cmd/server/admin/create_nodeconfig.go]: needs gofmt

File: pkg/cmd/server/api/v1/types.go [hunk 3]
```diff
@@ -53,6 +53,11 @@ type NodeConfig struct {
 	// system, this value should be set to the upstream nameservers dnsmasq resolves with.
 	DNSNameservers []string `json:"dnsNameservers"`
 
+	// DNSUpstreamResolvConf is a path to a resolv.conf file that contains settings for an upstream server.
+	// Only the nameservers and port fields are used. The file must exist and parse correctly. It overrides
+	// nameservers in DNSNameservers if set.
+	DNSUpstreamResolvConf string
```
Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: needs json tag

Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: add this field to ResolveNodeConfigPaths

Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: why override and not supplement the DNSNameservers field?

File: pkg/cmd/server/api/types.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/cmd/server/api/types.go]: It's the right word though - it's the resolv conf that we defer to.  DNSRecursiveResolvConf is another option I guess.

File: pkg/cmd/server/api/v1/types.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/cmd/server/api/v1/types.go]: Can I guess.

**Comment by smarterclayton** [comment]
Comments addressed

File: pkg/cmd/server/kubernetes/node/node_config.go [hunk 4]
```diff
@@ -350,6 +352,17 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig, enableProxy, enable
...
 		}
 
+		// allow an upstream resolv.conf to override nameservers.
+		if len(options.DNSRecursiveResolvConf) > 0 {
+			c, err := miekgdns.ClientConfigFromFile(options.DNSRecursiveResolvConf)
+			if err != nil {
+				return nil, fmt.Errorf("could not start DNS, unable to read config file: %v", err)
+			}
+			for _, s := range c.Servers {
+				dnsConfig.Nameservers = append(dnsConfig.Nameservers, net.JoinHostPort(s, c.Port))
```
Comment (liggitt) [inline pkg/cmd/server/kubernetes/node/node_config.go]: I expected this to combine with `options.DNSNameservers` if specified, but override the defaults we get from `dns.NewServerDefaults()`

File: pkg/cmd/server/api/types.go [hunk 5]
```diff
@@ -194,6 +194,11 @@ type NodeConfig struct {
 	// system, this value should be set to the upstream nameservers dnsmasq resolves with.
 	DNSNameservers []string
 
+	// DNSRecursiveResolvConf is a path to a resolv.conf file that contains settings for an upstream server.
+	// Only the nameservers and port fields are used. The file must exist and parse correctly. It overrides
+	// nameservers in DNSNameservers if set.
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: fix comment about overriding

**Comment by liggitt** [comment]
one comment on override behavior, LGTM otherwise

**Comment by smarterclayton** [comment]
Updated.

**Comment by smarterclayton** [comment]
Integration test hung / crashed

[test]

**Comment by smarterclayton** [comment]
[merge][severity: blocker]

We need dns on the node enabled and this blocks cluster lifecycle.

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

```
=== BEGIN TEST CASE ===
test/end-to-end/core.sh:214: executing 'docker pull 172.30.1.1:5000/test/hello-world:pullthrough' expecting success
FAILURE after 0.548s: test/end-to-end/core.sh:214: executing 'docker pull 172.30.1.1:5000/test/hello-world:pullthrough' expecting success: the command returned the wrong error code
Standard output from the command:
Trying to pull repository 172.30.1.1:5000/test/hello-world ... 
pullthrough: Pulling from 172.30.1.1:5000/test/hello-world
78445dd45222: Pulling fs layer

Standard error from the command:
error pulling image configuration: Get http://172.30.1.1:5000/v2/test/hello-world/blobs/sha256:48b5124b2768d2b917edcb640435044a97967015485e812545546cbed5cf0233: EOF
=== END TEST CASE ===
```

**Comment by smarterclayton** [comment]
[test] deployment config scale to zero flaked

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Force merging, known flake.

**Pull request merged**